pkg/rbd/controllerserver.go

```diff
@@ -525,104 +466,50 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 	if len(snapshotID) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "Snapshot ID cannot be empty")
 	}
-	snapshotIDMutex.LockKey(snapshotID)
 
+	snapshotIDMutex.LockKey(snapshotID)
 	defer func() {
 		if err := snapshotIDMutex.UnlockKey(snapshotID); err != nil {
 			klog.Warningf("failed to unlock mutex snapshot:%s %v", snapshotID, err)
 		}
 	}()
 
 	rbdSnap := &rbdSnapshot{}
-	if err := cs.MetadataStore.Get(snapshotID, rbdSnap); err != nil {
-		if err, ok := err.(*util.CacheEntryNotFound); ok {
-			klog.V(3).Infof("metadata for snapshot %s not found, assuming the snapshot to be already deleted (%v)", snapshotID, err)
-			return &csi.DeleteSnapshotResponse{}, nil
+	if err := genSnapFromSnapID(rbdSnap, snapshotID, req.GetSecrets()); err != nil {
+		// Consider missing snap as already deleted, and proceed to remove the omap values
+		if _, ok := err.(ErrSnapNotFound); !ok {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-
-		return nil, err
+		if err := unreserveSnap(rbdSnap, req.GetSecrets()); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.DeleteSnapshotResponse{}, nil
 	}
 
+	// lock out parallel create requests against the same snap name as we
+	// cleanup the image and associated omaps for the same
+	snapshotNameMutex.LockKey(rbdSnap.RequestName)
```

Comment:
Isnt this handled in ceph ?  do we need to serialize request here in CSI ?

Reply:
Parallel PVC creation rates should not be affected by this additional mutex.

The lock contention issue really depends on the granularity of mutexes in use. As we use a set of mutexes and as pointed out in the issue #279 we need to increase the number of mutexes that we can use to address the initial lock contention problem.

Reply:
The mutex against the RequestName, which is the SnapshotName as sent by the CO system, is required to ensure the RADOS omaps are protected against parallel updates from a creation path when being accessed by the delete path.

Creation, stores a `key` with the passed in VolumeName (or RequestName in this context) with the image UUID as its `value` (in the CSI directory omap). The next step is to update the image UUID directory (omap) to store a key with the RequestName (the back pointer). Finally the image with the said UUID is created. Hence, this is a 3 step process.

Deletion operates in the reverse, deleting the image first, then the image UUIDDirectory (omap), and finally removing the VolumeName key from the CSI directory omap.

Hence in order to protect against out of order updates to the above objects and keys, for a given RequestName, the name based lock is acquired.

This is critical and **cannot** be circumvented currently.

Performance concerns occur **only** when the mutex is contended, and that can happen if the name based hashed mutex collides (or SnapshotNames collide (VolumeNames should not as the CO system should protect against the same as per the specification)). This can be alleviated by increasing the number of mutexes as detailed in the issue #279.

Tagging @dillaman  to the comment to review reasoning above.

Reply:
I guess the bigger question is whether or not it's a legal operation for the CO to issue a delete before the create has completed? RBD doesn't currently offer any protection against racing creates/deletes since we've usually had a solid orchestration system above that would prevent things like that. About the closest issue I've had to deal with was a user w/ a home-grown orchestration system that resulted in racing image deletions against the same image. 

In a perfect world, the Ceph cluster would perform the directory updates atomically by performing two omap requests in a single atomic operation against the same object (that's what librbd does internally when storing a bi-directional name<->uuid mapping in the "rbd_directory" object). 

Another option would be to use Ceph locks (rados lock get), but once you start thinking about how to recover from a dead process holding a lock, things will start to get very complicated and ugly within ceph-csi.

[1] https://github.com/ceph/go-ceph

Reply:
On the legality of a Delete before Create, the spec has answers and it is logically not feasible to do so [1] as delete requires the volume ID which only a successful create can return.

But, there seems to be a possibility on a Create for a volume name that already exists [2] (but with different properties than what is being requested), which can race with a legal Delete of the existing volume. At which point the racy behavior may occur.

If the current locking granularity (as it is name hash based locks [3]) is the problem, then we may need to support granular locks based on ref counted lock objects created for each name/volID instance in flight. If that is acceptable then it can be the basis of a solution in the future?

[1] Request timeouts and negation: https://github.com/container-storage-interface/spec/blob/master/spec.md#timeouts

[2] Already existing but different properties for the existing volume: https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume-errors

[3] Kubernetes keymutex: https://github.com/kubernetes/utils/blob/master/keymutex/hashed.go

Reply:
@humblec the current thought process is that we move to more granular locks as a part of fixing #279 and hence the lock contention here and in other existing places is not an bottleneck in the future. Hope it addresses your concern.

Reply:
@ShyamsundarR the locking concerns and the race conditions I pointed out in our discussion still exist ( thanks for documenting some of it above on behalf) , however lets handle it differently. I am taking it in my todo list. for now, lets not worry about it in this PR .